### PR TITLE
[aarch64] Implement QGEMM kernels with UMMLA/SMMLA instructions

### PR DESCRIPTION
**Pull request by snadampal**
### Description

This PR adds UMMLA and SMMLA based QGEMM kernels for aarch64. This covers
(i) symmetric quantization (zero point is Zero)
(ii) asymmetric quantization (zero point is non zero)
(iii) per channel as well as per tensor quantization
(iv) Signed weights (U8S8 Gemm)
(v) Unsigned weights (U8U8 Gemm) and 
(vi) Signed activations and weights (S8S8 Gemm) scenarios

I've enabled the ummla/smmla kernels based on cpuinfo check for `I8MM` support 
MMLA QGEMM kernels are enabled for all the devices that support I8MM instructions.

### Motivation and Context

This is to improve INT8 quantized MatMul performance on aarch64 platform.
I have run the below benchmarking script (bert , roberta and gpt2 model inference) on AWS Graviton3 based c7g.4xl instance and observed up to 1.33x performance improvement compared to the optimized UDOT qgemm kernel performance.

```
cd onnxruntime/python/tools/transformers
python3 benchmark.py
```
I have also run the unit tests, and made sure all are passing

```
./build.sh --config RelWithDebInfo --build_shared_lib --parallel --compile_no_warning_as_error --skip_submodule_sync 

```

### CODE REVIEW

**Comment by snadampal** [comment]
Hi @snnn , would you be able to review and provide feedback on this PR? appreciate your time.

**Comment by milpuz01** [comment]
I have checked out the changes and run performance test and accuracy tests with and without flag using onnxruntime_perf_test (modified the binary to dump output for comparisons) on AWS Graviton3 instances and I can confirm that the performance uptick is as expected and accuracy was the same.

File: onnxruntime/core/mlas/lib/platform.cpp [hunk 1]
```diff
@@ -171,6 +171,14 @@ MlasInitAMX()
 
 #endif // MLAS_TARGET_AMD64_IX86
 
+static inline bool ort_use_int8mm() {
```
Comment (chenfucn) [inline onnxruntime/core/mlas/lib/platform.cpp:174]: consider changing it to a compilation flag?

**Review by chenfucn — COMMENTED** [review]
Windows assembly files are in here:
https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/core/mlas/lib/arm64

Also please make sure mlas unit tests pass when your new kernels are enabled.

File: onnxruntime/core/mlas/lib/platform.cpp (hunk 1 above)
Comment (snadampal) [reply onnxruntime/core/mlas/lib/platform.cpp:174]: Hi @chenfucn , do you mean make it compile time flag and enable the feature by default if the HW supports the I8MM instructions?

File: onnxruntime/core/mlas/lib/platform.cpp [hunk 2]
```diff
@@ -464,6 +472,22 @@ Return Value:
...
     }
 
+    //
+    // Check if the processor supports ASIMD I8MM instructions.
+    //
+
+    bool HasI8MMInstruction = false;
+
+    uint64_t isar1_el1;
+    asm("mrs %[reg], ID_AA64ISAR1_EL1\n" : [reg] "=r"(isar1_el1) : :);
```
Comment (skottmckay) [inline onnxruntime/core/mlas/lib/platform.cpp:482]: Reading ID_AA64ISAR1_EL1 will definitely cause a crash on an ARMv8-A chip and potentially cause a crash on later ARMv8 chips. See https://github.com/microsoft/onnxruntime/issues/17647 which we believe is due to reading that register.

Additionally I believe EL1 registers requires additional privileges to be read safely: https://github.com/golang/go/issues/57336#issuecomment-1353800393

Any reason we can't use the value from cpuinfo which is [read using sysctl](https://github.com/pytorch/cpuinfo/blob/959002f82d7962a473d8bf301845f2af720e0aa4/src/arm/mach/init.c#L381-L384)? Would need to expose it via CPUIDInfo as we do with other values like [HasArmNeoDot](https://github.com/microsoft/onnxruntime/blob/60f19ab001b40d8e0bd2912645afba28d2c016ba/onnxruntime/core/common/cpuid_info.h#L30) which is set using cpuinfo [here](https://github.com/microsoft/onnxruntime/blob/60f19ab001b40d8e0bd2912645afba28d2c016ba/onnxruntime/core/common/cpuid_info.cc#L139).

Comment (snadampal) [reply onnxruntime/core/mlas/lib/platform.cpp:482]: Hi @skottmckay , thanks for the review, I agree, in fact I started cleaning this up to use the library query, will post the updated version.

**Comment by snadampal** [comment]
Hi @chenfucn, @skottmckay , I have addressed your feedback and tested it on Linux. Could you please help trigger CI runs so that I can get the windows results and can fix if any issues. Thank you!

**Comment by chenfucn** [comment]
please add unit tests.

**Review by chenfucn — DISMISSED** [review]

**Comment by chenfucn** [comment]
hi @skottmckay this would increase binary size. Should we allow an exception or exclude this from minimal build?

**Comment by skottmckay** [comment]
> hi @skottmckay this would increase binary size. Should we allow an exception or exclude this from minimal build?

I think the benefit justifies any binary size increase, so I think it's fine to include this in a minimal build.

**Comment by skottmckay** [comment]
> Thanks for approving the PR, @chenfucn Chen Fu FTE . Hi @skottmckay Scott McKay FTE , the performance gains are substantial (up to 1.33x at use case level) to consider this feature even for a minimal build configuration. please let me know the next steps to get it merged. Thank you!

Need to resolve the build errors from the minimal builds. They're not due to any binary size checks.

e.g.
```
FAILED: CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/aarch64/QgemmU8X8KernelUmmla.S.o 
/ndk_home/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -target aarch64-none-linux-android29 --sysroot=/ndk_home/toolchains/llvm/prebuilt/linux-x86_64/sysroot -DCPUINFO_SUPPORTED_PLATFORM=1 -DDISABLE_FLOAT8_TYPES -DDISABLE_ML_OPS -DEIGEN_MPL2_ONLY -DEIGEN_USE_THREADS -DJSON_NOEXCEPTION -DMLAS_NO_EXCEPTION -DNSYNC_ATOMIC_CPP11 -DONNX_NO_EXCEPTIONS -DORT_EXTENDED_MINIMAL_BUILD -DORT_MINIMAL_BUILD -DORT_NO_EXCEPTIONS -DORT_NO_RTTI -DPLATFORM_POSIX -DUSE_NNAPI=1 -D_GNU_SOURCE -I/build/7/MinSizeRel/_deps/utf8_range-src -I/onnxruntime_src/include/onnxruntime -I/onnxruntime_src/include/onnxruntime/core/session -I/build/7/MinSizeRel/_deps/pytorch_cpuinfo-src/include -I/build/7/MinSizeRel/_deps/google_nsync-src/public -I/build/7/MinSizeRel -I/onnxruntime_src/onnxruntime -I/build/7/MinSizeRel/_deps/abseil_cpp-src -I/onnxruntime_src/onnxruntime/core/mlas/inc -I/onnxruntime_src/onnxruntime/core/mlas/lib -I/build/7/MinSizeRel/_deps/gsl-src/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -Os -DNDEBUG -fPIC -Wall -Wextra -Wno-unused-parameter -Werror  -march=armv8.2-a+i8mm -MD -MT CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/aarch64/QgemmU8X8KernelUmmla.S.o -MF CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/aarch64/QgemmU8X8KernelUmmla.S.o.d -o CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/aarch64/QgemmU8X8KernelUmmla.S.o -c /onnxruntime_src/onnxruntime/core/mlas/lib/aarch64/QgemmU8X8KernelUmmla.S
<instantiation>:45:18: error: invalid operand for instruction
        mov x27, v18.2D[0]
                 ^
<instantiation>:2:9: note: while in macro instantiation
        OutputRow12Element Zero,x2,x13,16,17,18,19,(8 == 1)
        ^
<instantiation>:25:9: note: while in macro instantiation
        OutputBlock Zero,12,8
        ^
<instantiation>:34:9: note: while in macro instantiation
        ProcessRows Zero,8
        ^
```

```
FAILED: CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp.o 
/ndk_home/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=aarch64-none-linux-android29 --sysroot=/ndk_home/toolchains/llvm/prebuilt/linux-x86_64/sysroot -DCPUINFO_SUPPORTED_PLATFORM=1 -DDISABLE_FLOAT8_TYPES -DDISABLE_ML_OPS -DEIGEN_MPL2_ONLY -DEIGEN_USE_THREADS -DJSON_NOEXCEPTION -DMLAS_NO_EXCEPTION -DNSYNC_ATOMIC_CPP11 -DONNX_NO_EXCEPTIONS -DORT_EXTENDED_MINIMAL_BUILD -DORT_MINIMAL_BUILD -DORT_NO_EXCEPTIONS -DORT_NO_RTTI -DPLATFORM_POSIX -DUSE_NNAPI=1 -D_GNU_SOURCE -I/build/7/MinSizeRel/_deps/utf8_range-src -I/onnxruntime_src/include/onnxruntime -I/onnxruntime_src/include/onnxruntime/core/session -I/build/7/MinSizeRel/_deps/pytorch_cpuinfo-src/include -I/build/7/MinSizeRel/_deps/google_nsync-src/public -I/build/7/MinSizeRel -I/onnxruntime_src/onnxruntime -I/build/7/MinSizeRel/_deps/abseil_cpp-src -I/onnxruntime_src/onnxruntime/core/mlas/inc -I/onnxruntime_src/onnxruntime/core/mlas/lib -I/build/7/MinSizeRel/_deps/gsl-src/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -ffunction-sections -fdata-sections -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -DCPUINFO_SUPPORTED -Os -DNDEBUG -std=gnu++17 -fPIC -fno-rtti -Wall -Wextra -Wno-unused-parameter -Wno-deprecated-copy -Wno-tautological-pointer-compare -Wno-ambiguous-reversed-operator -Wno-deprecated-anon-enum-enum-conversion -Wno-undefined-var-template -Wshorten-64-to-32 -Werror -MD -MT CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp.o -MF CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp.o.d -o CMakeFiles/onnxruntime_mlas.dir/onnxruntime_src/onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp.o -c /onnxruntime_src/onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp
/onnxruntime_src/onnxruntime/core/mlas/lib/qgemm_kernel_ummla.cpp:796:17: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
        int k = CountK;
```

and the lint warnings

https://github.com/microsoft/onnxruntime/actions/runs/6501885026/job/17685114761?pr=17160

**Comment by snadampal** [comment]
thanks, @skottmckay , I had tried the minimal build on Linux system but didn't observe any issues. Could you please share the  steps to reproduce the above build errors.
btw, I used the following command from this page: https://onnxruntime.ai/docs/build/custom.html#linux

```
<ONNX Runtime repository root>/build.sh \
  --config=Release \
  --build_shared_lib \
  --minimal_build \
  --disable_ml_ops --disable_exceptions --disable_rtti \
  --include_ops_by_config <config file from model conversion> --enable_reduced_operator_type_support \
  --skip_tests
```

I will fix the lint warnings.

**Comment by skottmckay** [comment]
It's not actually the minimal build aspect that's failing. Seems to be any Android build, which involves using the Android NDK to build (which uses clang). 

Are the new kernels expected to build for Android when the target architecture is arm64-v8a?

**Comment by snadampal** [comment]
I don't see a reason why they wouldn't, will give it a try. Are these the instructions for Android cross compilation?https://onnxruntime.ai/docs/build/android.html#cross-compiling-on-linux-and-macos

**Comment by skottmckay** [comment]
Yeah. Unfortunately requires installing the Android SDK and an NDK. Quickest and simplest is to install Android Studio and point to the SDK/NDK in that install. The API level doesn't matter too much - I used `29` in my build.

**Comment by snadampal** [comment]
I'm able to set up the Android sdk/ndk using sdkmanager from command line tools and reproduced the compilation failures. I will update the PR with the fixes.

**Comment by skottmckay** [comment]
Awesome! Thank you.

FWIW XNNPACK has some i8mm kernels. Not sure if those could be used as a reference to see if there's any difference in how they are implemented in general.

**Comment by snadampal** [comment]
thanks, @skottmckay , I'm able to fix the android clang build failures, the issue was trivial instruction syntax errors only. next I'm checking what linter and format tools are used to fix the cpplint warnings.

**Comment by skottmckay** [comment]
https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md#linting should have details of setting up the linting tools.

**Comment by snadampal** [comment]
Hi @skottmckay, thanks! I have pushed the updated commits and verified them for (1) Linux native compilation for libraries and wheels (2) linux native minimal build and (3) android ndk build, cross compiled on x86.

Regarding the cpplint warnings, looks like the similar warnings exist on several existing files as well, and most of them are of low confidence score, so, i didn't address some of them. But I have run the following format tools and made sure no errors or warnings.
```
lintrunner -a
git-clang-format
```

**Comment by skottmckay** [comment]
I think you can ignore the cpplint warnings for mlas files. That library was always special-cased when it came to formatting rules.

**Comment by snadampal** [comment]
The Windows_CI (CUDA and TVM) build failures don't seem to be related to the PR. Is it already a known issue? or do I need to look at?

**Comment by skottmckay** [comment]
Unrelated and not required builds. Please ignore.

**Comment by snadampal** [comment]
thank you,  please approve if there is no other feedback.

**Review by chenfucn — APPROVED** [review]

**Comment by snadampal** [comment]
Hi @chenfucn, thanks for the approval. Appreciate if you could merge it.

**Comment by snadampal** [comment]
thanks for merging the PR, @skottmckay !

**Comment by snnn** [comment]
@skottmckay , "post merge" pipeline found an issue from this PR: 

https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1179709&view=results

**Comment by snadampal** [comment]
Looks like iOS build failure, I'm not sure how it's different from the CI build.

**Comment by skottmckay** [comment]
@chenfucn Do the new kernels need to be treated the same as the special-cased fp16 ones (done in [this](https://github.com/microsoft/onnxruntime/pull/15270) PR) because custom `-march` value is required?

**Comment by skottmckay** [comment]
This didn't show up in the CIs as we only ran the required ones, which did not include iOS builds. We should probably make that required so things like this don't get missed in the future.

**Comment by snadampal** [comment]
I realized that I had the sources under generic aarch64 but the ISA compiler flag settings only under (not APPLE). not sure if it would have compiled if I had set the I8MM flags for Apple builds as well.
Anyway, based on the fp16 [CR](https://github.com/microsoft/onnxruntime/pull/15270), looks like it wouldn't compile, hence I have raised this PR to exclude the mmla kernels from Apple.
https://github.com/microsoft/onnxruntime/pull/18082

**Comment by snadampal** [comment]
Hi @snnn , could you please check this https://github.com/microsoft/onnxruntime/pull/18082 before reverting the mmla kernels?

**Pull request merged**